File: backend/app/schemas/ai.py
"""
Schemas for AI-powered configuration assistance
"""
from typing import List, Dict, Optional, Any
from decimal import Decimal
from pydantic import BaseModel, Field, field_serializer

from app.schemas.team import TeamMemberCreate
from app.schemas.service import ServiceCreate
from app.schemas.cost import CostFixedCreate
from app.core.pydantic_config import DECIMAL_CONFIG


class OnboardingSuggestionRequest(BaseModel):
    """Request for AI-powered onboarding suggestions"""
    industry: str = Field(..., description="Industry type (e.g., 'Marketing Digital', 'Desarrollo Web')")
    region: str = Field(default="US", description="Region code (e.g., 'US', 'CO', 'MX')")
    currency: str = Field(default="USD", description="Primary currency")
    custom_context: Optional[str] = Field(None, description="Additional context about the business")


class OnboardingSuggestionResponse(BaseModel):
    """Response with AI-generated onboarding suggestions"""
    suggested_roles: List[TeamMemberCreate] = Field(default_factory=list, description="Suggested team members")
    suggested_services: List[ServiceCreate] = Field(default_factory=list, description="Suggested services")
    suggested_fixed_costs: List[CostFixedCreate] = Field(default_factory=list, description="Suggested fixed costs")
    confidence_scores: Dict[str, float] = Field(default_factory=dict, description="Confidence scores for each suggestion category")
    reasoning: Optional[str] = Field(None, description="AI reasoning for the suggestions")


class DocumentParseRequest(BaseModel):
    """Request for parsing unstructured document data"""
    text: str = Field(..., description="Text content from document (PDF, CSV, etc.)")
    document_type: Optional[str] = Field(None, description="Type of document: 'payroll', 'expenses', 'mixed'")


class DocumentParseResponse(BaseModel):
    """Response with parsed structured data from document"""
    team_members: List[TeamMemberCreate] = Field(default_factory=list, description="Extracted team members")
    fixed_costs: List[CostFixedCreate] = Field(default_factory=list, description="Extracted fixed costs")
    subscriptions: List[Dict] = Field(default_factory=list, description="Extracted subscriptions")
    confidence_scores: Dict[str, float] = Field(default_factory=dict, description="Confidence scores for extracted data")
    warnings: List[str] = Field(default_factory=list, description="Warnings about parsing issues")


class NaturalLanguageCommandRequest(BaseModel):
    """Request for processing natural language configuration commands"""
    command: str = Field(..., description="Natural language command (e.g., 'Añade un Senior Designer que gana 45k anuales')")
    context: Optional[Dict] = Field(None, description="Current configuration context")


class NaturalLanguageCommandResponse(BaseModel):
    """Response with structured action from natural language command"""
    action_type: str = Field(..., description="Type of action: 'add_team_member', 'add_service', 'add_fixed_cost', etc.")
    action_data: Dict = Field(..., description="Structured data for the action")
    confidence: float = Field(..., ge=0.0, le=1.0, description="Confidence score for the parsed command")
    requires_confirmation: bool = Field(default=True, description="Whether user confirmation is required")
    reasoning: Optional[str] = Field(None, description="AI reasoning for the parsed command")


class ExecutiveSummaryService(BaseModel):
    """Servicio para resumen ejecutivo"""
    service_id: int = Field(..., description="ID del servicio", gt=0)
    service_name: str = Field(..., description="Nombre del servicio", min_length=1)
    estimated_hours: Optional[float] = Field(None, description="Horas estimadas", ge=0)
    client_price: Decimal = Field(..., description="Precio al cliente", gt=0)
    
    @field_serializer('client_price')
    def serialize_decimal(self, value: Decimal) -> str:
        return str(value)
    
    model_config = DECIMAL_CONFIG


class ExecutiveSummaryRequest(BaseModel):
    """Request para generar resumen ejecutivo"""
    project_name: str = Field(..., description="Nombre del proyecto", min_length=1)
    client_name: str = Field(..., description="Nombre del cliente", min_length=1)
    client_sector: Optional[str] = Field(None, description="Sector del cliente (ej: 'Tecnología', 'Retail')")
    services: List[ExecutiveSummaryService] = Field(..., description="Lista de servicios incluidos", min_items=1)
    total_price: Decimal = Field(..., description="Precio total de la cotización", gt=0)
    currency: str = Field("USD", description="Moneda", min_length=3, max_length=3)
    language: str = Field("es", description="Idioma del resumen: 'es' o 'en'", pattern="^(es|en)$")
    
    @field_serializer('total_price')
    def serialize_decimal(self, value: Decimal) -> str:
        return str(value)
    
    model_config = DECIMAL_CONFIG


class ExecutiveSummaryResponse(BaseModel):
    """Response con resumen ejecutivo generado"""
    summary: str = Field(..., description="Resumen ejecutivo generado")
    provider: str = Field("openai", description="Proveedor de IA utilizado")
    usage: Optional[Dict[str, Any]] = Field(None, description="Información de uso de la API (tokens, costo estimado)")

File: frontend/src/lib/queries/ai.ts
/**
 * React Query hooks for AI-powered configuration assistance
 */
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiRequest } from '../api-client';
import { queryKeys } from './queryKeys';

// Types
export interface OnboardingSuggestionRequest {
  industry: string;
  region?: string;
  currency?: string;
  custom_context?: string;
}

export interface TeamMemberSuggestion {
  name: string;
  role: string;
  salary_monthly_brute: number;
  currency: string;
  billable_hours_per_week: number;
  is_active?: boolean;
}

export interface ServiceSuggestion {
  name: string;
  description?: string;
  default_margin_target: number;
  pricing_type: 'hourly' | 'fixed' | 'recurring' | 'project_value';
  is_active?: boolean;
}

export interface FixedCostSuggestion {
  name: string;
  amount_monthly: number;
  currency: string;
  category: string;
  description?: string;
}

export interface OnboardingSuggestionResponse {
  suggested_roles: TeamMemberSuggestion[];
  suggested_services: ServiceSuggestion[];
  suggested_fixed_costs: FixedCostSuggestion[];
  confidence_scores: {
    roles?: number;
    services?: number;
    costs?: number;
  };
  reasoning?: string;
}

/**
 * Hook to get AI-powered onboarding suggestions
 */
export function useAISuggestions() {
  return useMutation({
    mutationFn: async (request: OnboardingSuggestionRequest): Promise<OnboardingSuggestionResponse> => {
      const response = await apiRequest<OnboardingSuggestionResponse>('/ai/suggest-config', {
        method: 'POST',
        body: JSON.stringify(request),
      });

      if (response.error) {
        throw new Error(response.error);
      }

      return response.data!;
    },
    onError: (error) => {
      console.error('Error getting AI suggestions:', error);
    },
  });
}

/**
 * Hook to check AI service status
 */
export function useAIStatus() {
  return useQuery({
    queryKey: queryKeys.ai.status(),
    queryFn: async () => {
      const response = await apiRequest<{ available: boolean; message: string }>('/ai/status', {
        method: 'GET',
      });

      if (response.error) {
        throw new Error(response.error);
      }

      return response.data!;
    },
    staleTime: 5 * 60 * 1000, // 5 minutes
  });
}

// Types for document parsing
export interface DocumentParseRequest {
  text: string;
  document_type?: 'payroll' | 'expenses' | 'mixed';
}

export interface ParsedTeamMember {
  name: string;
  role: string;
  salary_monthly_brute: string; // Decimal as string from API
  currency: string;
  billable_hours_per_week: number;
  is_active?: boolean;
}

export interface ParsedFixedCost {
  name: string;
  amount_monthly: string; // Decimal as string from API
  currency: string;
  category: string;
  description?: string;
}

export interface ParsedSubscription {
  name: string;
  amount_monthly: number;
  currency: string;
}

export interface DocumentParseResponse {
  team_members: ParsedTeamMember[];
  fixed_costs: ParsedFixedCost[];
  subscriptions: ParsedSubscription[];
  confidence_scores: {
    team_members?: number;
    fixed_costs?: number;
    subscriptions?: number;
  };
  warnings: string[];
}

/**
 * Hook to parse unstructured document data
 */
export function useParseDocument() {
  return useMutation({
    mutationFn: async (request: DocumentParseRequest): Promise<DocumentParseResponse> => {
      const response = await apiRequest<DocumentParseResponse>('/ai/parse-document', {
        method: 'POST',
        body: JSON.stringify(request),
      });

      if (response.error) {
        throw new Error(response.error);
      }

      return response.data!;
    },
    onError: (error) => {
      console.error('Error parsing document:', error);
    },
  });
}




